### ISSUE TITLE
libcamera: does not build reproducibly

### ISSUE THREAD

**Issue by raboof** [NixOS/nixpkgs#230289]
Building this package twice does not produce the bit-by-bit identical result each time, making it harder to detect CI breaches. You can read more about this at https://reproducible-builds.org/ .

Fixing bit-by-bit reproducibility also has additional advantages, such as avoiding hard-to-reproduce bugs, making content-addressed storage more effective and reducing rebuilds in such systems.

### Steps To Reproduce

```
nix-build '<nixpkgs>' -A libcamera --check --keep-failed
```

You can use `diffoscope` to analyze the differences in the output of the two builds.

To view the build log of the build that produced the artifact in the binary cache:

```
nix-store --read-log $(nix-instantiate '<nixpkgs>' -A libcamera)
```

**Comment by raboof** [comment]
It looks like this is somewhat intentional. If I understand things correctly, libcamera:

* generates a keypair during build
* signs the `ipa_ipu3` module with the private key (there used to be other modules, such as `ipa_rpi`, not sure what happened to those)
* embeds the public key in the library
* throws away the private key

Then at run time, it runs modules with valid signatures 'embedded', while other modules are 'isolated'.

It sounds like the intent is that it should be possible to have additional signed modules, beyond those shipped with libcamera. However, because we throw away the key after building, this is not possible in the Nix case. We might want to double-check whether we want to make this possible.

For the current simple case, a reproducible alternative to achieve the same result would be to not check signatures, but instead embed a list of hashes of trusted modules in the library. Modules that match one of the embedded hashes can be trusted, and those hashes can be reproducible. This might be something to contribute upstream.

But perhaps first we should update libcamera in nixpkgs to a more recent version, i.e. #258094 and futher to 0.2.0

**Comment by jpds** [comment]
0.2.0 at #281755

**Comment by raboof** [comment]
I have a possible solution to contribute upstream at https://codeberg.org/raboof/libcamera/commit/6ed45e756eed4dc1420352ab5547bb020d619d44

**Comment by kbingham** [comment]
The intention is to allow additional IPA modules which are *not* signed.

Those will be loaded as an isolated process. The signing process tells us if the module is open source compatible / allowed to be run in the same process space as libcamera - or if it is external and supplied by an untrusted vendor - thus must be run through a separate process and an IPC mechanism.

**Comment by kbingham** [comment]
Aha, indeed - so that's correct - but in /every/ case (not just nix) it is not possible to have additional signed modules beyond those signed by (the distribution).

It's our expectation that the distribution is in control of the signing. Either the default, as here by discarding the keys, or by ensuring that any trusted IPAs are signed accordingly.

In regards to your proposal on the libcamera-devel mailing list - I'll add that to our next libcamera call tomorrow to bring it up and get feedback.

**Comment by raboof** [comment]
> I'll add that to our next libcamera call tomorrow to bring it up and get feedback.

Thanks, much appreciated! Happy to make adjustments to that patch if the general approach seems sound :+1:

**Comment by kbingham** [comment]
Feedback from Laurent was, (I think he mentioned this on one of the patches too) - but an externally provided file containing the checksums isn't going to work. We can't make it 'easy' for vendors to circumvent the whole restrction in the first place by telling users "Here, put this binary blob in /usr/lib/xxx.so and add `03cfd743661f07975fa2f1220c5194cbaff48451 /usr/lib/xxx.so` to /usr/lib/libcamera.checksums.file

**Comment by raboof** [comment]
> Feedback from Laurent was, (I think he mentioned this on one of the patches too) - but an externally provided file containing the checksums isn't going to work. We can't make it 'easy' for vendors to circumvent the whole restrction in the first place by telling users "Here, put this binary blob in /usr/lib/xxx.so and add `03cfd743661f07975fa2f1220c5194cbaff48451 /usr/lib/xxx.so` to /usr/lib/libcamera.checksums.file

Right, that was the feedback on the first patch (https://lists.libcamera.org/pipermail/libcamera-devel/2024-January/040224.html) and fixed in the second patch (https://lists.libcamera.org/pipermail/libcamera-devel/2024-January/040244.html) which embeds the checksums in the binary instead of in an externally provided file.

**Comment by kbingham** [comment]
The same problem is there too.

$UntrustedVendor simply says "Please copy $badBinary to ${MESON_INSTALL_DESTDIR_PREFIX}/${badBinary} and run the officially sanctioned (provided by libcamera) src/ipa/ipa-checksum-install.sh script to circumvent the restrictions.

It's definitely a difficult problem :-( The key issue is /we/ should not supply the means that makes it easy for $UntrustedVendor to circumvent any restriction we put in place.

It's too close to leaving the key under the mat of the front door ...

And needing to link the modules to libcamera, makes it very tricky to add the checksums to libcamera with the linker directly.

**Comment by raboof** [comment]
Thanks for the feedback! 

> The same problem is there too.
> 
> $UntrustedVendor simply says "Please copy `$badBinary` to `${MESON_INSTALL_DESTDIR_PREFIX}/${badBinary}` and run the officially sanctioned (provided by libcamera) `src/ipa/ipa-checksum-install.sh` script to circumvent the restrictions.

Right - we wouldn't include `ipa-checksum-install.sh` in the binary package, so this would also require saying to the user "ok now download the libcamera sources and get the ipa-checksum-install.sh from there".

> It's definitely a difficult problem :-( The key issue is /we/ should not supply the means that makes it easy for $UntrustedVendor to circumvent any restriction we put in place.
> 
> It's too close to leaving the key under the mat of the front door ...
> 
> And needing to link the modules to libcamera, makes it very tricky to add the checksums to libcamera with the linker directly.

Of course libcamera being open source, there is no *entirely* foolproof mechanism possible (taken to extremes $UntrustedVendor can just tell people to use their fork of libcamera, or patch the binary to not perform the signature check...) - but we can see how far we can stretch making it hard for them.

A simple improvement might be to give the script a name that makes it more obvious that this script is meant as an implementation detail of the build process and not something users should be invoking directly (`libcamera-internal-ipa-checksum-install.sh`?), and perhaps require `--this-is-insecure-when-performed-outside-of-the-build --i-know-what-i-am-doing` flags. Would that work?

Beyond that, perhaps we could see if we can somehow embed the script into the meson build in a way that makes it impossible to invoke it outside of the build. Does that seem necessary?

**Comment by raboof** [comment]
Removing this issue from the ISO milestone because of the workaround in #307719, but keeping it open to track fixing this 'properly'.

**Comment by CertainLach** [comment]
In case of nix, we can see that modules are part of the same derivation, and trust them based on $out prefix, instead of trying to do something with signatures and hashes

Will this solve both reproducibility and trust issues?
Signatures won't work anyway, because if we use a static private key, then $UntrustedVendor can just... Use this static private key
